packages/freezed/lib/src/freezed_generator.dart

```diff
@@ -248,26 +240,48 @@ Read here: https://github.com/rrousselGit/freezed/blob/master/packages/freezed/C
   }
 
   List<Parameter> _commonParametersBetweenAllConstructors(
-    List<ConstructorDetails> constructorsNeedsGeneration,
-  ) {
+    List<ConstructorDetails> constructorsNeedsGeneration, {
+    required bool allowCommonSuperType,
+  }) {
     return constructorsNeedsGeneration.first.parameters.allParameters
         .map((parameter) {
-          var hasAnyFinalProperty = false;
-          for (final constructor in constructorsNeedsGeneration) {
-            final matchingParameter =
-                constructor.parameters.allParameters.firstWhereOrNull((p) {
-              return p.name == parameter.name && p.type == parameter.type;
-            });
+          var anyMatchingPropertyIsFinal = false;
+          var isCommonWithDifferentNullability = false;
+
+          DartType? commonType;
+
+          for (final constructor in constructorsNeedsGeneration.skip(1)) {
+            final matchingParameter = constructor.parameters.allParameters
+                .firstWhereOrNull((p) => p.name == parameter.name);
 
             if (matchingParameter == null) return null;
-            if (matchingParameter.isFinal) hasAnyFinalProperty = true;
-          }
 
-          if (hasAnyFinalProperty) {
-            return parameter.copyWith(isFinal: true);
+            final commonTypeWithMatch =
+                parameter.parameterElement!.library!.typeSystem.leastUpperBound(
+              parameter.parameterElement!.type,
+              matchingParameter.parameterElement!.type,
+            );
+
+            if (commonTypeWithMatch.isDartCoreObject) return null;
```

Comment:
Why return early here? Object is more specific than Object? (it gives you a non-null object, but I guess there isn't much you can do with Object). I would assume that if someone named all the parameters the same there could be a case made for wanting access to that parameter even if it returns Object?

Reply:
That is correct I think there should be an additional check to see if the original type was `Object`. The early return is there because Object is the supertype of everything so all "incompatible" types would be Object in the base mixin which in my opinion wouldn't be desirable.